src/MobileRedirectManager.php

```diff
@@ -0,0 +1,305 @@
+<?php
+/**
+ * Class MobileRedirect.
+ *
+ * @package AmpProject\AmpWP
+ */
+
+namespace AmpProject\AmpWP;
+
+use AMP_Options_Manager;
+
+/**
+ * Centralized manager to handle mobile redirection of a page.
+ *
+ * @package AmpProject\AmpWP
+ */
+final class MobileRedirectManager {
+
+	/**
+	 * The name of the cookie that persists the user's preference for viewing the non-AMP version of a page when on mobile.
+	 *
+	 * @var string
+	 */
+	const DISABLED_COOKIE_NAME = 'amp_mobile_redirect_disabled';
+
+	/**
+	 * Query parameter to indicate that the page in question should not be served as AMP.
+	 *
+	 * @var string
+	 */
+	const NO_AMP_QUERY_VAR = 'noamp';
+
+	/**
+	 * Indicates whether the the `amp_mobile_redirect_disabled` cookie has been set during the current request.
+	 *
+	 * When a cookie is set, it cannot be accessed via the `$_COOKIE` array until the next page load. This circumvent thia,
+	 * whenever the cookie is set during the current request, this variable becomes true and can be used as a fallback to
+	 * detect whether redirection has been disabled for the session.
+	 *
+	 * @var bool
+	 */
+	private static $disabled_cookie_is_set = false;
+
+	/**
+	 * Get whether mobile redirection is enabled or not.
+	 *
+	 * @return bool If JS redirection is disabled, only the status of the mobile redirection option is returned.
+	 *              Otherwise, returns true if mobile redirection option is enabled and current request is from a mobile device.
+	 */
+	public static function is_enabled() {
+		if ( ! AMP_Options_Manager::get_option( Option::MOBILE_REDIRECT ) ) {
+			return false;
+		}
+		return self::should_redirect_via_js() || self::is_mobile_request();
+	}
+
+	/**
+	 * Determine if mobile redirection is available for the current request.
+	 *
+	 * @return bool True if available, false otherwise.
+	 */
+	public static function is_available_for_request() {
+		return self::is_enabled() && is_amp_available();
+	}
+
+	/**
+	 * Determine if the current request is from a mobile device.
+	 *
+	 * @return bool True if current request is from a mobile device, otherwise false.
+	 */
+	public static function is_mobile_request() {
+		/**
+		 * Filters whether the current request is from a mobile device. This is provided as a means to short-circuit
+		 * the normal determination of a mobile request below.
+		 *
+		 * @since 1.6
+		 *
+		 * @param null $is_mobile Whether the current request is from a mobile device.
```

Comment:
In this case wouldn't the parameter always be null?

Reply:
In the case of filter docs, I think the phpdoc needs to capture what the filtered value _could_ be, not just what it initially is. Compare with the `pre_option_{$option}` filter. It is documented as:

```
	 * @param bool|mixed $pre_option The value to return instead of the option value. This differs from
	 *                               `$default`, which is used as the fallback value in the event the option
	 *                               doesn't exist elsewhere in get_option(). Default false (to skip past the
	 *                               short-circuit).
```

Reply:
OK, thanks for the clarification :+1: .